modules/mcuboot/CMakeLists.txt

```diff
@@ -299,10 +299,12 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
   endif ()
 
   foreach (filepath ${mcuboot_CONF_FILE})
-    file(STRINGS ${filepath} mcuboot_CONFIG_BOOT_SIGNATURE_KEY_FILE
-         REGEX "^CONFIG_BOOT_SIGNATURE_KEY_FILE=")
-    if (mcuboot_CONFIG_BOOT_SIGNATURE_KEY_FILE)
-      get_filename_component(mcuboot_CONF_DIR ${filepath} DIRECTORY)
+    if (EXISTS ${filepath})
+      file(STRINGS ${filepath} mcuboot_CONFIG_BOOT_SIGNATURE_KEY_FILE
+           REGEX "^CONFIG_BOOT_SIGNATURE_KEY_FILE=")
+      if (mcuboot_CONFIG_BOOT_SIGNATURE_KEY_FILE)
+        get_filename_component(mcuboot_CONF_DIR ${filepath} DIRECTORY)
+      endif()
     endif()
   endforeach()
```

Comment:
I really don't even see what the point of this code is, try it yourself, put `CONFIG_BOOT_SIGNATURE_KEY_FILE="<path>/bootloader/mcuboot/root-ec-p384.pem"` in a file `/tmp/my.conf` then use `west build -b nrf52840dk_nrf52840 zephyr/samples/hello_world -- -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_SECURE_BOOT=y -Dmcuboot_CONF_FILE="prj_minimal.conf;/tmp/my.conf" -Dmcuboot_OVERLAY_CONFIG=external_crypto.conf` and check where it's used:
```
grep -r "root-ec-p384.pem" build/
build/mcuboot/zephyr/include/generated/autoconf.h:#define CONFIG_BOOT_SIGNATURE_KEY_FILE "/tmp/aa/bootloader/mcuboot/root-ec-p384.pem"
build/mcuboot/zephyr/.config:CONFIG_BOOT_SIGNATURE_KEY_FILE="/tmp/aa/bootloader/mcuboot/root-ec-p384.pem"
build/mcuboot/shared_vars.cmake:set_property(TARGET mcuboot_shared_property_target  PROPERTY SIGNATURE_KEY_FILE;/tmp/aa/bootloader/mcuboot/root-ec-p384.pem)
build/mcuboot/build.ninja:build zephyr/autogen-pubkey.c | ${cmake_ninja_workdir}zephyr/autogen-pubkey.c: CUSTOM_COMMAND /tmp/aa/bootloader/mcuboot/root-ec-p384.pem || modules/nrf/ncs_version_h zephyr/app_version_h zephyr/driver_validation_h_target zephyr/kobj_types_h_target zephyr/offsets zephyr/offsets_h zephyr/parse_syscalls_target zephyr/syscall_list_h_target zephyr/version_h zephyr/zephyr_generated_headers
build/mcuboot/build.ninja:  COMMAND = cd /tmp/bb/build/mcuboot && /usr/bin/python /tmp/bb/bootloader/mcuboot/scripts/imgtool.py getpub -k /tmp/aa/bootloader/mcuboot/root-ec-p384.pem > /tmp/bb/build/mcuboot/zephyr/autogen-pubkey.c
build/build.ninja:  COMMAND = cd /tmp/bb/build/modules/mcuboot && /usr/bin/python /tmp/bb/bootloader/mcuboot/scripts/imgtool.py sign --key /tmp/aa/bootloader/mcuboot/root-ec-p384.pem --header-size 0x200 --align 4 --version 0.0.0+0 --pad-header --slot-size 0x77000 /tmp/bb/build/zephyr/mcuboot_primary_app.hex /tmp/bb/build/zephyr/app_signed.hex
build/build.ninja:  COMMAND = cd /tmp/bb/build/modules/mcuboot && /usr/bin/python /tmp/bb/bootloader/mcuboot/scripts/imgtool.py sign --key /tmp/aa/bootloader/mcuboot/root-ec-p384.pem --header-size 0x200 --align 4 --version 0.0.0+0 --pad-header --slot-size 0x77000 --pad /tmp/bb/build/zephyr/mcuboot_primary_app.hex /tmp/bb/build/zephyr/app_test_update.hex
build/build.ninja:  COMMAND = cd /tmp/bb/build/modules/mcuboot && /opt/zephyr-sdk-0.16.5-rc1/arm-zephyr-eabi/bin/arm-zephyr-eabi-objcopy --input-target=ihex --output-target=binary --gap-fill=0xff /tmp/bb/build/zephyr/mcuboot_primary_app.hex /tmp/bb/build/zephyr/app_to_sign.bin && /usr/bin/python /tmp/bb/bootloader/mcuboot/scripts/imgtool.py sign --key /tmp/aa/bootloader/mcuboot/root-ec-p384.pem --header-size 0x200 --align 4 --version 0.0.0+0 --pad-header --slot-size 0x77000 /tmp/bb/build/zephyr/app_to_sign.bin /tmp/bb/build/zephyr/app_update.bin
```
Then do the same but move the `my.conf` file to `<path>/bootloader/mcuboot/boot/zephyr` and give it a relative path `rm -rf build && west build -b nrf52840dk_nrf52840 zephyr/samples/hello_world -- -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_SECURE_BOOT=y -Dmcuboot_CONF_FILE="prj_minimal.conf;my.conf" -Dmcuboot_OVERLAY_CONFIG=external_crypto.conf`:
```
cd build && grep -r "root-ec-p384.pem" .
./mcuboot/zephyr/include/generated/autoconf.h:#define CONFIG_BOOT_SIGNATURE_KEY_FILE "/tmp/aa/bootloader/mcuboot/root-ec-p384.pem"
./mcuboot/zephyr/.config:CONFIG_BOOT_SIGNATURE_KEY_FILE="/tmp/aa/bootloader/mcuboot/root-ec-p384.pem"
./mcuboot/shared_vars.cmake:set_property(TARGET mcuboot_shared_property_target  PROPERTY SIGNATURE_KEY_FILE;/tmp/aa/bootloader/mcuboot/root-ec-p384.pem)
./mcuboot/build.ninja:build zephyr/autogen-pubkey.c | ${cmake_ninja_workdir}zephyr/autogen-pubkey.c: CUSTOM_COMMAND /tmp/aa/bootloader/mcuboot/root-ec-p384.pem || modules/nrf/ncs_version_h zephyr/app_version_h zephyr/driver_validation_h_target zephyr/kobj_types_h_target zephyr/offsets zephyr/offsets_h zephyr/parse_syscalls_target zephyr/syscall_list_h_target zephyr/version_h zephyr/zephyr_generated_headers
./mcuboot/build.ninja:  COMMAND = cd /tmp/bb/build/mcuboot && /usr/bin/python /tmp/bb/bootloader/mcuboot/scripts/imgtool.py getpub -k /tmp/aa/bootloader/mcuboot/root-ec-p384.pem > /tmp/bb/build/mcuboot/zephyr/autogen-pubkey.c
./build.ninja:  COMMAND = cd /tmp/bb/build/modules/mcuboot && /usr/bin/python /tmp/bb/bootloader/mcuboot/scripts/imgtool.py sign --key /tmp/aa/bootloader/mcuboot/root-ec-p384.pem --header-size 0x200 --align 4 --version 0.0.0+0 --pad-header --slot-size 0x77000 /tmp/bb/build/zephyr/mcuboot_primary_app.hex /tmp/bb/build/zephyr/app_signed.hex
./build.ninja:  COMMAND = cd /tmp/bb/build/modules/mcuboot && /usr/bin/python /tmp/bb/bootloader/mcuboot/scripts/imgtool.py sign --key /tmp/aa/bootloader/mcuboot/root-ec-p384.pem --header-size 0x200 --align 4 --version 0.0.0+0 --pad-header --slot-size 0x77000 --pad /tmp/bb/build/zephyr/mcuboot_primary_app.hex /tmp/bb/build/zephyr/app_test_update.hex
./build.ninja:  COMMAND = cd /tmp/bb/build/modules/mcuboot && /opt/zephyr-sdk-0.16.5-rc1/arm-zephyr-eabi/bin/arm-zephyr-eabi-objcopy --input-target=ihex --output-target=binary --gap-fill=0xff /tmp/bb/build/zephyr/mcuboot_primary_app.hex /tmp/bb/build/zephyr/app_to_sign.bin && /usr/bin/python /tmp/bb/bootloader/mcuboot/scripts/imgtool.py sign --key /tmp/aa/bootloader/mcuboot/root-ec-p384.pem --header-size 0x200 --align 4 --version 0.0.0+0 --pad-header --slot-size 0x77000 /tmp/bb/build/zephyr/app_to_sign.bin /tmp/bb/build/zephyr/app_update.bin
```

And for confirmation with debug added to the code here, the first invocation:
```
CMake Warning at /tmp/bb/nrf/modules/mcuboot/CMakeLists.txt:302 (message):
  check prj_minimal.conf


CMake Warning at /tmp/bb/nrf/modules/mcuboot/CMakeLists.txt:302 (message):
  check my.conf


CMake Warning at /tmp/bb/nrf/cmake/partition_manager.cmake:79 (message):
  

          ---------------------------------------------------------------------
          --- WARNING: Using a bootloader without pm_static.yml.            ---
```
And second:
```
CMake Warning at /tmp/bb/nrf/modules/mcuboot/CMakeLists.txt:302 (message):
  check prj_minimal.conf


CMake Warning at /tmp/bb/nrf/modules/mcuboot/CMakeLists.txt:302 (message):
  check /tmp/my.conf


CMake Warning at /tmp/bb/nrf/modules/mcuboot/CMakeLists.txt:307 (message):
  used!
```

Reply:
I admit that solution is not the best one and it could be improved (e.g. by using `APPLICATION_CONFIG_DIR` as base for relative path here). My intention was to avoid breaking backwards compatibility here and just add a hotfix. Better solution is more than welcome. :)

This code works for case where MCUboot configuration is redefined in main application configuration directory (without passing the `mcuboot_CONF_FILE` through a CMake option). Then CMake here receives an absolute path. Actually that's how some applications define their MCUboot key now (e.g. nRF Desktop).

Reply:
Have you got a build string for an application that does this?

Reply:
`nrf/applications/nrf_desktop`; `west build -b nrf52820dongle_nrf52820 -p -- -DCONF_FILE=prj_release.conf`

Reply:
I've just ran that and it doesn't do anything with MCUboot, it configures b0, s0 and s1